### PR TITLE
FIX: allow time and timestamp without time zone in AT TIME ZONE

### PR DESCRIPTION
**Pull request by geraint0923**
Allow time and timestamp without time zone in AT TIME ZONE

### CODE REVIEW

**Comment by electrum** [comment]
Don't put "FIX" in commit messages. Format like this:

"Allow time and timestamp without time zone in AT TIME ZONE"

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -747,12 +747,19 @@ protected Type visitAtTimeZone(AtTimeZone node, StackableAstVisitorContext<Analy
         {
             Type valueType = process(node.getValue(), context);
             process(node.getTimeZone(), context);
-            if (!valueType.equals(TIME_WITH_TIME_ZONE) && !valueType.equals(TIMESTAMP_WITH_TIME_ZONE)) {
-                throw new SemanticException(TYPE_MISMATCH, node.getValue(), "Type of value must be a time or timestamp with time zone (actual %s)", valueType);
+            if (!valueType.equals(TIME_WITH_TIME_ZONE) && !valueType.equals(TIMESTAMP_WITH_TIME_ZONE) && !valueType.equals(TIME) && !valueType.equals(TIMESTAMP)) {
+                throw new SemanticException(TYPE_MISMATCH, node.getValue(), "Type of value must be a time or timestamp with/without time zone (actual %s)", valueType);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Make this message "with or without"

**Comment by electrum** [comment]
Please add a test for this.

**Comment by martint** [comment]
> Please add a test for this.

This got caught by the product tests.

**Comment by ilfrin** [comment]
as much as I am a fan of having automated product tests and I advocated for running them here, I think if there's a possibility to test something on w lower level we should. 

Not sure if that product test was aimed specifically at that issue. Maybe so, then we're fine. In general case I'd leave product tests to serve more as smoke, testing the "glue" and interactions between components and proving something works end-to-end for real. If they caught this that's great, but I wouldn't rely on them to catch every detail and exercise all flows. Integration and unit tests are faster to execute and easier to spot the actual locate errors.

**Comment by martint** [comment]
Yeah, I didn't mean to imply that we shouldn't add the test (especially because it makes it easier to test from the IDE or command line), but I was just pointing out that it wasn't a fix being done in a vacuum.

**Comment by martint** [comment]
I'm just glad that the product tests were able to catch the bug :)

**Comment by ilfrin** [comment]
Cool, i get it :)

Yeah we're also happy that they proved useful, on the first day

usually at the beginning I'd expect a lot of false negatives, so our first
response was: fix the test :)

czw., 17.03.2016, 19:26 użytkownik Martin Traverso <notifications@github.com>
napisał:

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java [hunk 2]
```diff
@@ -21,14 +23,39 @@
...
+import static com.facebook.presto.spi.type.TimeType.TIME;
+import static com.facebook.presto.spi.type.TimeWithTimeZoneType.TIME_WITH_TIME_ZONE;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.facebook.presto.spi.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+import static java.util.Objects.requireNonNull;
+
 public class DesugaringRewriter
         extends ExpressionRewriter<Void>
 {
+    IdentityHashMap<Expression, Type> expressionTypes;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java]: Make this `private final`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java [hunk 3]
```diff
@@ -21,14 +23,39 @@
...
+    {
+        this.expressionTypes = requireNonNull(expressionTypes, "expressionTypes is null");
+    }
+
     @Override
     public Expression rewriteAtTimeZone(AtTimeZone node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
     {
+        Expression value = treeRewriter.rewrite(node.getValue(), context);
+        Type valueType = expressionTypes.get(value);
+        Expression coercedValue = value;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java]: I'd get rid of this variable and do this, instead:

```java
Expression value = ...
Type type = ...

if (type.equals(TIME)) {
    value = new Cast(value, ...);
}
else if (type.equals(TIMESTAMP)) {
    value = new Cast(value, ...);
}

return new FunctionCall(...);
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java [hunk 4]
```diff
@@ -21,14 +23,39 @@
...
+    public DesugaringRewriter(IdentityHashMap<Expression, Type> expressionTypes)
+    {
+        this.expressionTypes = requireNonNull(expressionTypes, "expressionTypes is null");
+    }
+
     @Override
     public Expression rewriteAtTimeZone(AtTimeZone node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
     {
+        Expression value = treeRewriter.rewrite(node.getValue(), context);
+        Type valueType = expressionTypes.get(value);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/DesugaringRewriter.java]: Call this `type`

**Comment by martint** [comment]
Add a comment to the commit explaining
- what was broken
- why you had to move the tests

**Comment by geraint0923** [comment]
@martint 
Addressed the comments and updated the commit message.

**Comment by martint** [comment]
Merged, thanks!

**Pull request closed without merging**